src/common/lib/python/common_extension.cc

```diff
@@ -104,6 +108,8 @@ PyObject *PyTask_from_string(PyObject *self, PyObject *args) {
   result = (PyTask *) PyObject_Init((PyObject *) result, &PyTaskType);
   result->size = size;
   result->spec = TaskSpec_copy((TaskSpec *) data, size);
+  /* The created task does not include any execution dependencies. */
+  result->execution_dependencies = empty_execution_dependencies;
```

Comment:
Is this used because the `PyTask` is created with `malloc` and not `new`? Why not just say `result->execution_dependencies = std::vector<ObjectID>();`?

Reply:
Yeah, I think because the `PyTask` is created with `malloc`, setting it to `std::vector<ObjectID>()` tries to deallocate the previous value of `result->execution_dependencies`, causing a memory error. I'm not sure why this version works, actually... It also worked on my machine if you set a temporary variable like `x= std::vector<ObjectID>()` and then set `result->execution_dependencies = x`.

Reply:
@mehrdadn @pcmoritz do you know what the best way to do this is? We have a C++ struct which contains a `std::vector`, which is defined at https://github.com/ray-project/ray/pull/1281/files#diff-1dac80a71f6c9a5afd7428d1f0635674R28. And it is allocated with `malloc`. We can't control this because it is used to define a Python object, so the `malloc` is called from within `PyObject_New` at https://github.com/ray-project/ray/pull/1281/files#diff-1feaf7f51d81690b43f5c06a58034f82R107

Reply:
@robertnishihara @pcmoritz @stephanie-wang Ah, yeah, you can't assign to an object that isn't constructed. What you need is a "placement new" operator, i.e. to construct an object in existing memory. The syntax is
```
void *address = malloc(sizeof(T));
T *ptr = new(address) T(...);
// ...
ptr->~T();
free(address);
```
Normally you should avoid that operator when possible, but this is one of the very rare instances where you can (indeed, must) use that operator. :)

Oh, and by the way, in cases where you _do_ need to be assigning, you don't need `empty_execution_dependencies`. It doesn't save you any efficiency. There's no heap allocation for an empty vector, so you should just do `execution_dependencies.clear()` or `execution_dependencies = std::vector<ObjectID>()` (if you want the capacity to remain the same), or `std::vector<ObjectID>().swap(execution_dependencies)` (if you want the underlying memory to be freed).

Reply:
Interesting, so in this case we'd probably need to store `ptr` as a field inside of the `PyTask` struct, which is pretty weird, but maybe necessary. Another workaround would be to just have the `PyTask` contain a `std::vector<ObjectID> *` instead of `std::vector<ObjectID>`. Not sure if there would be a performance hit there.

Reply:
@robertnishihara Oh sorry, forgot to mention that as far as I know the posters are always the same, so you can just cast. I just posted the usual syntax.

Reply:
For now, I'll just store a `std::vector<ObjectID> *`. I don't think it's necessary for us to construct the vector from the existing memory. If it turns out to have a performance hit, we can often set it to `NULL` when there are no execution dependencies.